src/lib/data/tob.ts

```diff
@@ -0,0 +1,649 @@
+import { calcPercentOfNum, calcWhatPercent, percentChance, randInt, reduceNumByPercent, round, Time } from 'e';
+import { KlasaUser } from 'klasa';
+import { Bank } from 'oldschooljs';
+import { SkillsEnum } from 'oldschooljs/dist/constants';
+
+import { checkUserCanUseDegradeableItem } from '../degradeableItems';
+import { constructGearSetup, GearStats } from '../gear';
+import { blowpipeDarts } from '../minions/functions/blowpipeCommand';
+import { getMinigameScore } from '../settings/minigames';
+import { UserSettings } from '../settings/types/UserSettings';
+import { Gear } from '../structures/Gear';
+import { Skills } from '../types';
+import { assert, formatSkillRequirements, randomVariation, skillsMeetRequirements } from '../util';
+import getOSItem from '../util/getOSItem';
+import resolveItems from '../util/resolveItems';
+
+export const bareMinStats: Skills = {
+	attack: 90,
+	strength: 90,
+	defence: 90,
+	ranged: 90,
+	magic: 94,
+	prayer: 77
+};
+
+export const totalXPFromRaid = {
+	[SkillsEnum.Attack]: 12_000,
+	[SkillsEnum.Hitpoints]: 13_100,
+	[SkillsEnum.Strength]: 12_000,
+	[SkillsEnum.Ranged]: 1000,
+	[SkillsEnum.Defence]: 12_000,
+	[SkillsEnum.Magic]: 1000
+} as const;
+
+export interface TOBRoom {
+	name: string;
+	difficultyRating: number;
+	timeWeighting: number;
+}
+export const TOBRooms: TOBRoom[] = [
+	{
+		name: 'Maiden',
+		difficultyRating: 1,
+		timeWeighting: 11
+	},
+	{
+		name: 'Bloat',
+		difficultyRating: 4,
+		timeWeighting: 11
+	},
+	{
+		name: 'Nylocas',
+		difficultyRating: 3,
+		timeWeighting: 22
+	},
+	{
+		name: 'Sotetseg',
+		difficultyRating: 2,
+		timeWeighting: 11
+	},
+	{
+		name: 'Xarpus',
+		difficultyRating: 2,
+		timeWeighting: 17
+	},
+	{
+		name: 'Verzik',
+		difficultyRating: 6,
+		timeWeighting: 28
+	}
+];
+
+interface TOBDeaths {
+	/**
+	 * An array, where each item is the index of the room they died in.
+	 */
+	deaths: number[];
+	deathChances: { name: string; deathChance: number }[];
+}
+
+export function calculateTOBDeaths(
+	kc: number,
+	hardKC: number,
+	attempts: number,
+	hardAttempts: number,
+	isHardMode: boolean,
+	gear: {
+		melee: number;
+		range: number;
+		mage: number;
+		total: number;
+	}
+): TOBDeaths {
+	let deaths: number[] = [];
+	let deathChances: { name: string; deathChance: number }[] = [];
+
+	// This shifts the graph left or right, to start getting kc sooner or later. Higher = sooner:
+	const minionLearningBase = isHardMode ? 5 : 2;
+	const minionLearning = minionLearningBase + Math.floor(Math.min(20, (isHardMode ? hardAttempts : attempts) / 2));
+	const basePosition = isHardMode ? 92 : 85;
+	const difficultySlider = isHardMode ? 32 : 36; // Lower is harder. Be careful with this. (30 default).
+	const curveStrength = isHardMode ? 2 : 2.5; // 1 - 5. Higher numbers mean a slower learning curve. (2.5 default)
+	const baseKC = isHardMode ? hardKC : kc;
+
+	let baseDeathChance = Math.floor(
```

Comment:
Can replace Math.sqrt(100) with 10. 

That's on me, obviously, but I left it  as sqrt(100) during development to make it easier to tweak.

```ts
let baseDeathChance = Math.floor(
		basePosition - (Math.log(baseKC / curveStrength + minionLearning) / Math.log(10)) * difficultySlider
	);
```


Maybe leave it for now in case we need to do more balancing?